P5-217: Added more about RTS (#139)
aau-hcl-p5/main: report/5.Solution/Scheduling.tex

Keep\_USB\_Alive    & 14                            \\
Receive\_Data       & 13                            \\
Move\_Motors        & 28                            \\
Handle\_Lasers      & 30                            \\
Update\_Display     & 480                           \\                     
\end{tabular}
\end{table}\label{table:executionTimes}

Since the execution time of all the tasks totals to a mere 535 microseconds, it was deemed non-essential to include minor cycles, as the tasks all depend on receiving new data from the USB, and this will not occur more than once every millisecond due to hardware limitations.
Likewise, there is a software limitation when using nxtOSEK, since the interrupt service routine is only run once every millisecond.
With these considerations, the cyclic execution method with equal minor and major cycles were deemed the best solution, since the implementation of a fixed priority task would add more overhead when changing between tasks and preempting them, rather than just allowing the previous tasks to finish executing.

\subsection{Fixed Priority Scheduling}

> However, as the project progressed, parts of the machine intelligence execution was moved to the NXT for execution.

Reviewer: ```suggestion
However as the project progressed, parts of the machine intelligence execution was moved to the NXT for execution.
```

Reviewer: det er én ledsætnings i begyndelsen -> Kun ét komma

Author: 😶 nej nej nej spang

This lead to tasks that took longer than 1 millisecond to execute, rendering the use of cyclic executive infeasible, as described in Subsection~\cite{Design:Scheduling}. 

> Due to this, the fixed-priority scheduling model was instead utilized. 

Reviewer 2: ```suggestion
Instead, the fixed-priority scheduling model was utilized. 
```

> After having added machine learning to the \texttt{Move\_Motors} task the execution times are as follows:

Reviewer 2: Føler måske at det her skal rykkes lidt op, men det kan vi lige tage mandag. Tror det giver mere mening at bruge det har som argument for hvorfor vi ikke bruger CE, men hey vi kan sagtens bruge FPS.

\begin{table}[tbp]
\begin{tabular}{lll}
\textbf{Task name}  & \textbf{Execution time} 	& \textbf{Priority}\\
Keep\_USB\_Alive    & 14                        & 6    \\
Receive\_Data       & 13                        & 4   \\
Move\_Motors        & --                        & 2   \\
Handle\_Lasers      & 30                        & 3   \\
Update\_Display     & 480                       & 1   \\                     
Background\_Task    & -                        	& 5  \\                     
\end{tabular}
\end{table}\label{table:executionTimes}
\todo{Add execution time of move motors and bg task}

> As the cyclic executive approach allowed running the calibration before entering the first major cycle, this was easily implemented.

Reviewer 2: ```suggestion
Cyclic executive allowed for execution of tasks, before the major cycles.
This was used for running the calibration, which needed to be run only once, and before the other tasks, .
```

Author: den her er ikke helt korrekt, den tager vi lige mandag

der er mere end en major cycle, så det skal før den første af dem

> However, this does not work with the fixed-priority scheduling model, as the tasks are repeatedly run according to their tasks.

Reviewer 2: Hjælp

To fix this, the \texttt{Background\_Task} was introduced, which has a high enough priority to preempt the other tasks on the system to allow calibration to run, while still keeping the USB connection alive.

\begin{lstlisting}[language=CSharp,label={lst:backgroundTask},caption={Background task}]
TASK(background_task) {
    if(!calibrated){
        show_init_screen();
        while(!get_status_code(&current_status, 0));

        if (current_status == READY_FOR_CALIBRATION) {
            calibrate(false);
        }
        calibrated = true;
    }
    if (current_status == DISCONNECTED_REQ) {
        stop();
    }
    TerminateTask();
}
\end{lstlisting}

> As seen on Snippet~\ref{lst:backgroundTask}, the task checks if the boolean value calibrated has been changed to true, meaning that the calibration has already be done.

Reviewer 2: Jeg kiggede faktisk: https://ell.stackexchange.com/questions/120749/on-listing-or-in-listing
```suggestion
As seen in Snippet~\ref{lst:backgroundTask}, the task checks if the boolean value, calibrated, is true.
Calibrated indicates if calibration is completed.
```

If this is not the case, the background task ensures that the status of the motors and USB are set as disconnected using the \texttt{stop} function.

> In the cyclic executive model, it was ensured that the \texttt{keep\_USB\_alive} task was executed once every millisecond by running it at the beginning at each cycle with a cycle length of 1 millisecond.

Reviewer 2: ```suggestion
In the cyclic executive model, it was ensured that the \texttt{keep\_USB\_alive} task was executed once every millisecond by running it at the beginning of each cycle, as the cycles had a length of 1 millisecond.
```

> With the FPS model, this is instead handled by assigning the task the highest possible priority, ensuring that it will run regardless of whatever else is currently being executed using preemption if needed.

Reviewer 2: ```suggestion
With the FPS model, this is instead done by assigning the \texttt{keep\_USB\_alive} task with the highest possible priority.
This ensures that it will run ahead of all other tasks, using preemption if needed.
```

> To further ensure that it runs exactly once every 1 millisecond, an alarm is specified in the OIL file, as seen on Snippet~\ref{lst:OILAlarm}.

Reviewer 2: ```suggestion
To further ensure that it is run exactly once every 1 millisecond, an alarm is specified in the OIL file, as seen in Snippet~\ref{lst:OILAlarm}.
```

Reviewer 2: Føler jeg, da det vel egentligt er en deadline der skal overholdes?

\begin{lstlisting}[language=CSharp,label={lst:OILAlarm},caption={Alarm in the OIL}]
  /* Definition of keep_USB_alive execution timing */
  ALARM cyclic_keep_USB_alive
  {
    COUNTER = SysTimerCnt;
    ACTION = ACTIVATETASK
    {
      TASK = keep_USB_alive;
    };
    AUTOSTART = TRUE
    {
      ALARMTIME = 1;
      CYCLETIME = 1; /* keep_USB_alive is executed every msec */
      APPMODE = appmode1;
    };
  };
\end{lstlisting}

> The counter attribute specifies that the system clock is used to time when the alarm should be run, and which action it should take upon activation.

Reviewer 2: Counter skal nok have noget formatting, vist egentligt generelt for den her.

Reviewer 2: Also hjælp, forstår ikke denne sætning

> In the autostart block of the definition, it is defined that the cycletime is 1, meaning once every update of the SysTimerCnt, which takes place once every 1 millisecond.

Reviewer 2: ```suggestion
In the autostart block, it is defined that the cycletime is 1, meaning once every update of the SysTimerCnt, which takes place once every 1 millisecond.
```

Reviewer 2: Hjælp, den er snørklet.
